### ISSUE TITLE
Message-Authenticator is only included when doing EAP

### ISSUE THREAD

**Issue by arjanoosting** [radcli/radcli#107]
Hi,

Due to the BlastRadius vulnerability it is recommended to include the Message-Authenticator field in all Access-Request and Access-Reply packets. Radcli doesn't include them in the Access-Request unless doing EAP as can be seen in the code in lib/send-server.c.

I think something like this should work:

```
 @@ -672,8 +672,8 @@ int rc_send_server_ctx(rc_handle * rh, RC_AAA_CTX ** ctx, SEND_DATA * data,
 
                auth->length = htons((unsigned short)total_length);
 
-               /* If EAP message we MUST add a Message-Authenticator attribute */
-               if (rc_avpair_get(data->send_pairs, PW_EAP_MESSAGE, 0) != NULL)
+               /* Send an Message-Authenticator with all access requests and replies */
+               if (data->code == PW_ACCESS_REQUEST || data->code == PW_ACCESS_ACCEPT || data->code == PW_ACCESS_DENY)
                    total_length = add_msg_auth_attr(rh, secret, auth, total_length);
        }

```

but I am no C-programmer and don't know if this part of the code will actually be used to send Access replies as well.

I am also wondering if radcli does validate the Message-Authenticator field if it receives it in a packet?

**Comment by pem** [comment]
Yes, this needs to be addressed. However, if I read RFC 3579 correctly, it's only needed with Access Request.

From "3.2.  Message-Authenticator":

`This attribute MAY be used to authenticate and integrity-protect
      Access-Requests in order to prevent spoofing.  It MAY be used in
      any Access-Request.  It MUST be used in any Access-Request,
      Access-Accept, Access-Reject or Access-Challenge that includes an
      EAP-Message attribute.`

Customers are setting "require_message_authenticator = yes" (or equivalent) on the server end, which is effective regardless of the EAP attribute. If it's not set, it seems it just ignores any Message-Authenticator, so it should be safe to set it for all Access Requests.

**Comment by nmav** [comment]
Could you please verify that #111 addresses the issue?

**Comment by herwinw** [comment]
For the full description of the underlying issue, see https://www.blastradius.fail/ (especially the in depth description and the draft RFC "Deprecating Insecure Practices in RADIUS"
The simplified version is "Include Message-Authenticator in the packets you send, require and validate Message-Authenticator in the packets you receive"

The change in #111 fixes the first issue, by always adding the Message-Authenticator to the Access-Request packets.
To complement this, you want to be able to verify the reply of the server includes the Message-Authenticator attribute (this can be done with `rc_avpair_get`, I'm not sure if an more specific function to make this easier to check is really necessary). And in the case the server includes a Message-Authenticator attribute, you want to validate the contents of this attribute. I could not find anything in the code that performs this last part, it could probably be added to `rc_check_reply`.

These two parts are disjoint from each other, so IMHO #111 can be merged to fix this first issue, the second issues can be addressed in a separate change.

**Comment by nmav** [comment]
This is currently implemented and will be part of 1.5.0. A more advanced protection comes from the #113 pull request, that also makes sense to include.
